In [2]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt
from tensorflow.keras import layers
from tensorflow.keras.preprocessing.image import ImageDataGenerator

c:\Users\Raghunandan\anaconda3\Lib\site-packages\pandas\core\computation\expressions.py:22: UserWarning: Pandas requires version '2.10.2' or newer of 'numexpr' (version '2.10.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


In [3]:
dataset_path=r"D:\Tekworks\Deep-Learning(DL)\CNN\Cat_Dog_Dataset\PetImages"

In [4]:
datagen=ImageDataGenerator(
    # Normalize pixels
    rescale=1./255,
    #Data augmentation
    shear_range=0.2,zoom_range=0.2,horizontal_flip=True,
    #Split dataset for training and validation sets
    validation_split=0.2)

In [5]:
#Load training data
train_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='training'
)

Found 19978 images belonging to 2 classes.


In [6]:
val_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='validation'
)

Found 4993 images belonging to 2 classes.


In [7]:
print(train_data.class_indices)

{'Cat': 0, 'Dog': 1}


In [8]:
# Building CNN
model=keras.models.Sequential()

In [9]:
# First convolution layer
model.add(
    layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu',
        input_shape=(128,128,3)
    )
)

C:\Users\Raghunandan\AppData\Roaming\Python\Python313\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [10]:
# First pooling layer
model.add(
    layers.MaxPooling2D(
        pool_size=(2,2),
    )
)

In [11]:
# Second Convolution layer
model.add(
    layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu'
    )
)

In [12]:
# Second pooling layer
model.add(
    layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [13]:
# Third Convolution layer
model.add(
    layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu'
    )
)

In [14]:
# Third Pooling layer
model.add(
    layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [15]:
# Flattening layer
model.add(layers.Flatten())

In [16]:
# Dense layer
model.add(layers.Dense(128,activation='relu'))

In [17]:
# Output layer
model.add(layers.Dense(1,activation='sigmoid'))

In [18]:
# Model architecture
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 61, 61, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 30, 30, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 28, 28, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 14, 14, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 6272)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       802,944 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 822,465 (3.14 MB)

 Trainable params: 822,465 (3.14 MB)

 Non-trainable params: 0 (0.00 B)

#### First Convolution layer

#### (None,126,126,Channels)

#### (Batchsize,Height,Width,No.of filters)

Input = 128x128x3

filter = 3x3

Output shape Formula = ceil((N-F+2P)/S) + 1 (N-input image size, F-kernel size, P-padding, S-stride)

Output shape = ceil((128-3+0)/1)+1

Output shape = 126x126x32 

#### Params

Formula = ((kh * kw * input channels)+ 1) * filters

Kernel = 3x3

Input channels = 3

Filters = 32

Params = (3 * 3 * 3 + 1) * 32

Params = 896

#### First Pooling layer

MaxPooling2D((2,2))

Pooling = No learning, no weights (only reduces size)

Input = 126x126x32

pool = 2x2

Formula = Input image size/pool size --> (126/2)

Output size = 63

Final = (None,63,63,32)

#### Second convolution layer

Input from pooling : (None,63,63,32)

filter = 3x3

Output shape = ceil((63-3+0)/1)+1

Output shape = 61x61x32

#### Params

Input from convolution : (None,61,61,32)

Params = ((3 * 3 * 32)+1)*32

Params = 9248

#### Second pooling layer

Input = (None,61,61,32)

Pooling = 61/2

Output size = 30x30x32

#### Flattening layer

Input = 30x30x32

Output = 30x30x32

Output of flattening = 28800 (3D->1D)

#### Dense Layer

Input vector = 28800

Dense layer = (28800 x neurons in the layer)+biases

Dense layer = (28800x128)+128

Output = 36,86,528

#### Output layer

Input = 128

Output = 1

Output shape = (None,1)  --> Single Prediction (Cat or Dog)

Parameter = (128 * 1 )+ 1

Parameter = 129

Input --> Conv1 --> Pool1 --> Conv2 --> Pool2 --> Flattening --> Dense --> Output

In [19]:
import tensorflow as tf
print(tf.config.list_physical_devices('GPU'))

[]


In [20]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [21]:
history=model.fit(train_data,validation_data=val_data,epochs=2)

Epoch 1/2
360/625 ━━━━━━━━━━━━━━━━━━━━ 2:14 509ms/step - accuracy: 0.5277 - loss: 0.6982

C:\Users\Raghunandan\AppData\Roaming\Python\Python313\site-packages\PIL\TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


625/625 ━━━━━━━━━━━━━━━━━━━━ 422s 670ms/step - accuracy: 0.5941 - loss: 0.6608 - val_accuracy: 0.6607 - val_loss: 0.6216
Epoch 2/2
625/625 ━━━━━━━━━━━━━━━━━━━━ 116s 185ms/step - accuracy: 0.7227 - loss: 0.5482 - val_accuracy: 0.7537 - val_loss: 0.5012


In [22]:
loss,acc=model.evaluate(val_data)
print(f"Accuracy : {acc}")
print(f"Loss : {loss}")

157/157 ━━━━━━━━━━━━━━━━━━━━ 20s 129ms/step - accuracy: 0.7551 - loss: 0.5054
Accuracy : 0.7550570964813232
Loss : 0.5053626894950867


In [23]:
pred=model.predict(val_data)
pred_classes=(pred>0.5).astype(int)
image,label=next(train_data)
print("Original labels : ")
print(label[:5])
print("Predicted labels : ")
print(pred_classes[:5])

157/157 ━━━━━━━━━━━━━━━━━━━━ 20s 127ms/step
Original labels : 
[1. 1. 0. 0. 0.]
Predicted labels : 
[[1]
 [0]
 [1]
 [1]
 [0]]


In [24]:
model.save("cnn_cat_dog.h5")